### PR TITLE
Fix DAG to accommodate multiple inputs on initial node

### PR DESCRIPTION
**Pull request by swissrussdev**
Minor corrections to DAG

### CODE REVIEW

**Comment by swissrussdev** [comment]
The Unit Test was wrong : used a transformer which didn't match up to the input/output structures and names provided in the Action. I've corrected it and added a TestTransformer which replicates the required behaviour.

File: src/main/scala/io/smartdatalake/workflow/DAG.scala [hunk 1]
```diff
@@ -267,15 +268,18 @@ object DAG extends SmartDataLakeLogger {
...
+    val targetIDsforIncomingIDsMap = edges.groupBy(_.nodeIdFrom).mapValues(_.map(_.nodeIdTo))
+    nodes.map(n => (n, targetIDsforIncomingIDsMap.getOrElse(n.nodeId, Seq()))).toMap
   }
 
   /**
    * Create a lookup table to retrieve incoming (source) node IDs for a node.
    */
   private def buildIncomingIdLookupTable(nodes: Seq[DAGNode], edges: Seq[DAGEdge]) = {
-    val incomingIDsForTargetIDMap = edges.groupBy(_.nodeIdTo).mapValues(_.map(_.nodeIdFrom))
+    //val incomingIDsForTargetIDMap = edges.groupBy(_.nodeIdTo).mapValues(_.map(_.nodeIdFrom))
```
Comment (zzeekk) [inline src/main/scala/io/smartdatalake/workflow/DAG.scala:279]: H @swissrussdev , easier would be 
`val incomingIDsForTargetIDMap = edges.groupBy(_.nodeIdTo).mapValues(_.map(_.nodeIdFrom).distinct)
`

File: src/main/scala/io/smartdatalake/workflow/DAG.scala [hunk 2]
```diff
@@ -292,8 +296,9 @@ object DAG extends SmartDataLakeLogger {
     assert(startNodeIds.nonEmpty, s"Loop detected in remaining nodes ${incomingIds.keys.mkString(", ")}")
     // remove start nodes from incoming node list of remaining nodes
     val remainingNodeIdsWithoutIncomingStartNodes = remainingNodeIds.mapValues {
-      inIds => inIds.diff(startNodeIds.keys.toSeq)
+      inIds => inIds.filterNot(startNodeIds.keys.toSet) //eliminated all occurrences of seen nodes, not just the first one
```
Comment (zzeekk) [inline src/main/scala/io/smartdatalake/workflow/DAG.scala:299]: Nice, didn't no you can use a Set as predicate.

File: src/main/scala/io/smartdatalake/workflow/DAG.scala (hunk 1 above)
Comment (zzeekk) [reply src/main/scala/io/smartdatalake/workflow/DAG.scala]: Ok, i pushed the change. I'll merge then for release 1.0 :-)

**Comment by kaelte** [comment]
I will create a pull request with a tail recursive implementation of 

https://github.com/smart-data-lake/smart-data-lake/blob/6b3746f58e58ec1f3207cd45114ad8675ab63d8d/src/main/scala/io/smartdatalake/workflow/DAG.scala#L288

File: src/main/scala/io/smartdatalake/workflow/DAG.scala [hunk 3]
```diff
@@ -292,8 +296,9 @@ object DAG extends SmartDataLakeLogger {
     assert(startNodeIds.nonEmpty, s"Loop detected in remaining nodes ${incomingIds.keys.mkString(", ")}")
     // remove start nodes from incoming node list of remaining nodes
     val remainingNodeIdsWithoutIncomingStartNodes = remainingNodeIds.mapValues {
```
[the remark this answers is not in the document]
Comment (zzeekk) [reply src/main/scala/io/smartdatalake/workflow/DAG.scala:295]: Interesting. In favor of? 
For now where on Scala 2.11 and 2.12... so i wouldn't change this.

File: src/main/scala/io/smartdatalake/workflow/DAG.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (swissrussdev) [reply src/main/scala/io/smartdatalake/workflow/DAG.scala]: Excellent - thank you for all the improvements.

File: src/main/scala/io/smartdatalake/workflow/DAG.scala [hunk 4]
```diff
@@ -292,8 +296,9 @@ object DAG extends SmartDataLakeLogger {
     assert(startNodeIds.nonEmpty, s"Loop detected in remaining nodes ${incomingIds.keys.mkString(", ")}")
```
[the remark this answers is not in the document]
Comment (swissrussdev) [reply src/main/scala/io/smartdatalake/workflow/DAG.scala]: Making sortStep private appears appropriate here

File: src/main/scala/io/smartdatalake/workflow/DAG.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (swissrussdev) [reply src/main/scala/io/smartdatalake/workflow/DAG.scala]: Closing this as @kaelte retained mapValues in his PR

**Comment by swissrussdev** [comment]
I've incorporated the improvements from @kaelte - everything passes now and we've got some cleaner code into the bargain. Waiting for final confirm from @kaelte before merge.

**Comment by pgruetter** [comment]
Great, thanks a lot for the improvements. Build also seems to work again, so please merge this one as soon as you have the approval from@kaelte, then we can try to release 1.0.0 this week.

**Comment by swissrussdev** [comment]
Thanks everyone for your patience with me!

**Pull request merged**